### PR TITLE
Cecabank: Encrypt credit card fields

### PR DESCRIPTION
**Pull request by sinourain**
Description:
------------------------------
Encrypt credit_card and 3ds sensitive fields for Cecabank gateway. Also, I took the opportunity to do a small refactor in the identification network_transaction_id

For Spreedly reference:
[SER-929](https://spreedly.atlassian.net/browse/SER-929)

Remote Test:
------------------------------
Finished in 31.100887 seconds.
16 tests, 59 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed
0.51 tests/s, 1.90 assertions/s

Unit Tests:
------------------------------
Finished in 23.98564 seconds.
5748 tests, 78729 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed
239.64 tests/s, 3282.34 assertions/s

RuboCop:
------------------------------
784 files inspected, no offenses detected

### CODE REVIEW

**Review by javierpedrozaing — COMMENTED** [review]
LGTM

File: lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb [hunk 1]
```diff
@@ -51,34 +51,53 @@ def purchase(money, creditcard, options = {})
...
         before_message = before_message.gsub(%r(\")i, '\\\"')
         after_message = request_data.to_json.gsub(%r(\")i, '\\\"')
         transcript.sub(before_message, after_message)
       end
 
       private
 
+      def decrypt_sensible_fields(data)
+        cipher = OpenSSL::Cipher.new('AES-256-CBC').decrypt
+        cipher.key = [@options[:sensitive_fields_signature_key]].pack('H*')
```
Comment (Heavyblade) [inline lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb:88]: ❓ Question:

Wait, you are adding another GSF,  what's the difference between `sensitive_fields_signature_key` and `cypher_key`?

**Review by Heavyblade — COMMENTED** [review]
@sinourain done with the review, left some comments and a warning

File: lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb (hunk 1 above)
Comment (Heavyblade) [reply lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb:88]: Ahhh I can see the  why, 

in this case, the name for the GSF is not entirely correct this is an `encryption key` not a `signature key`,  something short like `encryption_key` should be enough

File: lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb [hunk 2]
```diff
@@ -51,34 +51,53 @@ def purchase(money, creditcard, options = {})
...
+        params['encryptedData'] = encrypt_sensible_fields(sensible_fields)
+        request_data['parametros'] = encode_params(params.to_json)
         before_message = before_message.gsub(%r(\")i, '\\\"')
         after_message = request_data.to_json.gsub(%r(\")i, '\\\"')
         transcript.sub(before_message, after_message)
       end
 
       private
 
+      def decrypt_sensible_fields(data)
```
[the remark this answers is not in the document]
Comment (naashton) [reply lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb:86]: @Heavyblade, thanks for this callout. Are we sure this fixed IV isn't just for the sandbox? Either way, we can just inject the IV from our end and set it from the `options` hash.

File: lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb (hunk 1 above)
Comment (naashton) [reply lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb:88]: Is the cipher key specific to the merchant or is there one cipher key that will be used for encrypting the required payload on behalf of all Cecabank users?

Comment (sinourain) [reply lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb]: Yes, `cipher.key` is specific for the gateway, it is different to the used for encrypting the payload

File: lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb [hunk 3]
```diff
@@ -51,34 +51,53 @@ def purchase(money, creditcard, options = {})
...
         request_data = JSON.parse(before_message)
-        params =  decode_params(request_data['parametros']).
-                  gsub(%r(("pan\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
-                  gsub(%r(("caducidad\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
-                  gsub(%r(("cvv2\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
-                  gsub(%r(("csc\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]')
-        request_data['parametros'] = encode_params(params)
+        params = parse(request_data['parametros'])
 
+        sensible_fields = decrypt_sensible_fields(params['encryptedData']).
```
[the remark this answers is not in the document]
Comment (curiousepic) [reply lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb:70]: `sensible` should be `sensitive`

**Comment by curiousepic** [comment]
`sensible` has a different meaning, should be `sensitive`

**Review by naashton — APPROVED** [review]
This looks good! Nice work @sinourain!

**Pull request merged**